cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -849,18 +877,93 @@
 (rule (cmp size src1 src2)
       (cmp_rmi_r size (CmpOpcode.Cmp) src1 src2))
 
+;; Helper for creating `MInst.XmmCmpRmR` instructions.
+(decl xmm_cmp_rm_r (SseOpcode RegMem Reg) ProducesFlags)
+(rule (xmm_cmp_rm_r opcode src1 src2)
+      (ProducesFlags.ProducesFlags (MInst.XmmCmpRmR opcode src1 src2)
+                                   (invalid_reg)))
+
+;; Helper for creating `fpcmp` instructions (cannot use `fcmp` as it is taken by
+;; `clif.isle`).
+(decl fpcmp (Value Value) ProducesFlags)
+(rule (fpcmp src1 @ (value_type $F32) src2)
+      (xmm_cmp_rm_r (SseOpcode.Ucomiss) (put_in_reg_mem src1) (put_in_reg src2)))
+(rule (fpcmp src1 @ (value_type $F64) src2)
+      (xmm_cmp_rm_r (SseOpcode.Ucomisd) (put_in_reg_mem src1) (put_in_reg src2)))
+
 ;; Helper for creating `test` instructions.
 (decl test (OperandSize RegMemImm Reg) ProducesFlags)
 (rule (test size src1 src2)
       (cmp_rmi_r size (CmpOpcode.Test) src1 src2))
 
-;; Helper for creating `MInst.Cmove` instructions.
+;; Helper for creating `cmove` instructions. Note that these instructions do not
+;; always result in a single emitted x86 instruction; e.g., XmmCmove uses jumps
+;; to conditionally move the selected value into an XMM register.
 (decl cmove (Type CC RegMem Reg) ConsumesFlags)
-(rule (cmove ty cc consequent alternative)
+(rule (cmove (is_gpr_type ty) cc consequent alternative)
       (let ((dst WritableReg (temp_writable_reg ty))
             (size OperandSize (operand_size_of_type_32_64 ty)))
         (ConsumesFlags.ConsumesFlags (MInst.Cmove size cc consequent alternative dst)
                                      (writable_reg_to_reg dst))))
+(rule (cmove (is_xmm_type ty) cc consequent alternative)
+      (let ((dst WritableReg (temp_writable_reg ty))
+            (size OperandSize (operand_size_of_type_32_64 ty)))
+        (ConsumesFlags.ConsumesFlags (MInst.XmmCmove size cc consequent alternative dst)
+                                     (writable_reg_to_reg dst))))
+
+;; Helper for creating `cmove` instructions directly from values. This allows us
+;; to special-case the `I128` types and default to the `cmove` helper otherwise.
+;; It also eliminates some `put_in_reg*` boilerplate in the lowering ISLE code.
+(decl cmove_from_values (Type CC Value Value) ConsumesFlags)
+(rule (cmove_from_values $I128 cc consequent alternative)
+      (let ((cons ValueRegs (put_in_regs consequent))
+            (alt ValueRegs (put_in_regs alternative))
+            ;; TODO these hard-coded sizes must go.
+            (dst1 WritableReg (temp_writable_reg $I64))
+            (dst2 WritableReg (temp_writable_reg $I64))
+            (size OperandSize (OperandSize.Size64))
+            ;; TODO there must be some way to say that both of these
+            ;; instructions consume the flags together.
+            (_ Unit (emit (MInst.Cmove size cc (RegMem.Reg (value_regs_get cons 0)) (value_regs_get alt 0) dst1))))
```

Comment:
Note that we will want a `with_flags_*` variant that takes the `SmallVec<[ConsumesFlags; 2]>` as well, and it will probably have to be implemented in external Rust code so it can loop over each `ConsumesFlags` in the small vec.

Reply:
This sounds like a lot of new infrastructure to build and this PR may already have enough of that. The current lowering rules perform the same unsafe emission as this PR. Can't this be done afterwards?

Reply:
The current lowering rules for `select` are written in Rust, so it isn't really comparable. It would be one thing if this `emit` was near the `emit` for the flags-producer instruction and the `emit` for the other flags-consuming instruction, but it isn't. It isn't even clear to me, reading this code, that the instructions will always be emitted in the correct order!

It seems to me that the order of evaluation will be something like

* construct `FlagsProducer` (do not emit it)
* call this constructor
  * emit first cmove
  * construct FlagsConsumer with the second cmove
* call a `with_flags` variant with the constructed `FlagsProducer` and `FlagsConsumer`
  * emit the producer
  * emit the consumer
 
If that is actually how things end up evaluating, then we would get `cmove; test/cmp; cmove` instead of `test/cmp; cmove; cmove`.

It looks like we aren't testing `select` with `i128`s yet (since you have only ported floating point `select`s so far). But this is exactly the kind bug I want to avoid with a dedicated `with_flags_*` variant that takes `SmallVec<[ConsumesFlags; 2]>`, as described above.

I really think we should fix this issue before landing this PR, or change this PR to not define and use `cmove_from_values` and instead use plain `cmove` (which will work for all the single-register values for `select` that are are currently supported in ISLE).

Reply:
Ok, @cfallin and I (but mostly @cfallin), worked on this a bit. The unpolished results are in [667935f](https://github.com/bytecodealliance/wasmtime/pull/3682/commits/667935f6ae12122be0454ba905c4c665c55404e1)--@fitzgen, thoughts?

Reply:
The extensions to `with_flags`, `ProducesFlags`, and `ConsumesFlags` look good!